ZohoCRMCharts/Classes/ZCRMFunnel.swift

```diff
@@ -0,0 +1,1148 @@
+	fileprivate let titleView: UILabel = UILabel()
+	fileprivate let conversionRateView: UILabel = UILabel()
+	fileprivate let stackView1: UIStackView = UIStackView()
+	fileprivate let stackView2: UIStackView = UIStackView()
+	fileprivate let collectionViewFlowLayout: UICollectionViewFlowLayout = UICollectionViewFlowLayout()
+	fileprivate var collectionView: UICollectionView!
+	fileprivate let standardConversionRateView: ZCRMStandardFunnelConversionRateView = ZCRMStandardFunnelConversionRateView()
+	fileprivate let barChart: ZCRMBarChart = ZCRMBarChart()
+	
+	public init(title: String, type: ZCRMCharts.ZCRMFunnelType, stages: [ZCRMFunnelStage]) {
+		self.title = title
+		self.type = type
+		self.stages = stages
+		super.init(frame: .zero)
+		self.render()
+	}
+	
+	public init(frame: CGRect, title: String, type: ZCRMCharts.ZCRMFunnelType, stages: [ZCRMFunnelStage]) {
+		self.title = title
+		self.type = type
+		self.stages = stages
+		super.init(frame: frame)
+		self.render()
+	}
+	
+	public init(title: String, type: ZCRMCharts.ZCRMFunnelType, stages: [ZCRMFunnelStage], segments: [ZCRMFunnelSegment]) {
+		self.title = title
+		self.type = type
+		self.stages = stages
+		self.segments = segments
+		super.init(frame: .zero)
+		self.render()
+	}
+	
+	public init(frame: CGRect, title: String, type: ZCRMCharts.ZCRMFunnelType, stages: [ZCRMFunnelStage], segments: [ZCRMFunnelSegment]) {
+		self.title = title
+		self.type = type
+		self.stages = stages
+		self.segments = segments
+		super.init(frame: frame)
+		self.render()
+	}
+	
+	required init?(coder aDecoder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+	
+	public override func layoutSubviews() {
+		self.deactivateConstraints()
+		self.addConstraints()
+	}
+	
+	public func reloadData() {
+		
+		if dataSource != nil {
+			self.loadData()
+			if self.type != .segment {
+				self.setData()
+			}
+		}
+	}
+}
+
+//MARK: - UI Rendering.
+
+fileprivate extension ZCRMFunnel{
+	
+	
+	fileprivate var rowCount: Int {
+		return self.segments.count + (self.type == .segment ? 2 : 1)
+	}
+	
+	func render() {
+		
+		self.clipsToBounds = true
+		self.renderTitleView()
+		self.renderView()
+	}
+	
+	func addConstraints() {
+		
+		if self.type == .path {
+			self.addPathFunnelConstraints()
+		} else if self.type == .compact {
+			self.addCompactFunnelConstraints()
+		} else if self.type == .classic {
+			self.addClassicFunnelConstraints()
+		} else if self.type == .segment {
+			self.addSegmentFunnelConstraints()
+		} else {
+			self.addStandardFunnelConstraints()
+		}
+	}
+	
+	func updateChanges() {
+		
+		switch self.type {
+		case .path:
+			self.updatePathUIOptions()
+		case .compact:
+			self.updateCompactFunnelUIOptions()
+		case .classic:
+			self.updateClassicFunnelUIOptions()
+		case .segment:
+			self.updateSegmentUIOptions()
+		case .standard:
+			self.updateStandardUIOptions()
+		}
+	}
+	
+	private func renderView() {
+		
+		if self.type == .path {
+			self.renderPathFunnel()
```

Comment:
use switch case.

Reply:
In swift both switch and if are doing the same thing.

Reply:
switch lead to code that is easier to read and debug, It also enables us to make our control flows more declarative. Swift is more powerful in pattern matching so we can use pattern matching instead of conditional checking.